File: GMGextractor/src/Utils/Logger.cs
using System;
using System.IO;

namespace GMGextractor.Utils
{
    public class Logger
    {
        private readonly string _logFilePath;
        private readonly string _errorLogFilePath;
        
        public Logger()
        {
            // Create logs directory in the application path
            string appPath = AppDomain.CurrentDomain.BaseDirectory;
            string logDir = Path.Combine(appPath, "logs");
            
            if (!Directory.Exists(logDir))
            {
                Directory.CreateDirectory(logDir);
            }
            
            // Set up log file paths
            _logFilePath = Path.Combine(logDir, "app.log");
            _errorLogFilePath = Path.Combine(logDir, "error.log");
        }
        
        /// <summary>
        /// Format a log message with timestamp
        /// </summary>
        private string FormatLogMessage(string level, string message)
        {
            return $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] [{level}] {message}";
        }
        
        /// <summary>
        /// Write log message to console and file
        /// </summary>
        private void Log(string level, string message, bool isError = false)
        {
            string formattedMessage = FormatLogMessage(level, message);
            
            // Log to console
            Console.WriteLine(formattedMessage);
            
            // Log to file
            try
            {
                File.AppendAllText(_logFilePath, formattedMessage + Environment.NewLine);
                
                // Also log errors to a separate file
                if (isError)
                {
                    File.AppendAllText(_errorLogFilePath, formattedMessage + Environment.NewLine);
                }
            }
            catch (Exception ex)
            {
                // If logging to file fails, just print to console
                Console.WriteLine($"Error writing to log file: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Log an informational message
        /// </summary>
        public void LogInfo(string message)
        {
            Log("INFO", message);
        }
        
        /// <summary>
        /// Log a warning message
        /// </summary>
        public void LogWarning(string message)
        {
            Log("WARNING", message);
        }
        
        /// <summary>
        /// Log an error message
        /// </summary>
        public void LogError(string message)
        {
            Log("ERROR", message, true);
        }
        
        /// <summary>
        /// Log a debug message (only in debug mode)
        /// </summary>
        public void LogDebug(string message)
        {
            #if DEBUG
            Log("DEBUG", message);
            #endif
        }
    }
}

File: backend/utils/logger.js
const fs = require('fs');
const path = require('path');

// Create logs directory if it doesn't exist
const logsDir = path.join(__dirname, '..', 'logs');
if (!fs.existsSync(logsDir)) {
  fs.mkdirSync(logsDir, { recursive: true });
}

// Log file paths
const logFile = path.join(logsDir, 'app.log');
const errorLogFile = path.join(logsDir, 'error.log');

/**
 * Logger utility for application
 */
class Logger {
  /**
   * Format a log message with timestamp
   * @param {string} level - Log level
   * @param {string} message - Log message
   * @returns {string} - Formatted log message
   */
  static formatLogMessage(level, message) {
    const timestamp = new Date().toISOString();
    return `[${timestamp}] [${level}] ${message}\n`;
  }
  
  /**
   * Write log message to console and file
   * @param {string} level - Log level
   * @param {string} message - Log message
   * @param {boolean} isError - Whether this is an error log
   */
  static log(level, message, isError = false) {
    const formattedMessage = this.formatLogMessage(level, message);
    
    // Log to console
    console.log(formattedMessage);
    
    // Log to file
    fs.appendFileSync(logFile, formattedMessage);
    
    // Log errors to separate error log file
    if (isError) {
      fs.appendFileSync(errorLogFile, formattedMessage);
    }
  }
  
  /**
   * Log info message
   * @param {string} message - Info message
   */
  static info(message) {
    this.log('INFO', message);
  }
  
  /**
   * Log warning message
   * @param {string} message - Warning message
   */
  static warn(message) {
    this.log('WARN', message);
  }
  
  /**
   * Log error message
   * @param {string} message - Error message
   */
  static error(message) {
    this.log('ERROR', message, true);
  }
  
  /**
   * Log debug message
   * @param {string} message - Debug message
   */
  static debug(message) {
    if (process.env.NODE_ENV !== 'production') {
      this.log('DEBUG', message);
    }
  }
}

module.exports = Logger;
